### PR TITLE
Support stdin journal (web), create journal file if not exists (ui, web)

### LINKED ISSUE #855: better hledger-web first startup
Reported on list today: hledger-web fails to start when no journal file exists yet. And if you have started it by clicking icons you won’t know why. We should provide a better startup experience.

### PR DESCRIPTION
**Pull request by zarybnicky**
Closes #855. The hledger-web half of #950.

This also fixes some oversights regarding `"-"` journal files in hledger-lib and
hledger (some, not all).

The behavior with `-f -` is currently the following: the journal is loaded into
memory, and any /add commands append to the in-memory journal. /edit doesn't do
anything yet as `readJournalFiles` uses `readFileOrStdinPortably` and assumes
the file is either `-` or a file, so there's some postprocessing (filters, anon,
...) that would need to be reimplemented.

The #950 part is just using `ensureJournalFileExists` in place of
`requireJournalFileExists`.

### CODE REVIEW

**Comment by zarybnicky** [comment]
I forgot to mention - it it possible to implement this without touching hledger-lib or hledger, but there would be a bit more special-casing required.

Looking back, I'm not sure about the `writeValidJournal` change, esp. if it shouldn't be a `pure ()` instead, but I was looking at `appendToJournalFileOrStdout` at that time...

**Comment by zarybnicky** [comment]
Completing /edit and /upload would also be a good step that I'd wanted to implement but ran out of time yesterday. I'm not sure if I should continue in this PR (and mark this one as WIP) or open another one later.

**Comment by zarybnicky** [comment]
Oh, and also - the exceptions ("handle is already closed") were caused by the check in `journalSpecifiedFileIsNewer`, not the initial loading of the file.

**Comment by simonmichael** [comment]
Thanks, I'll take a look asap.

File: hledger-web/Hledger/Web/Main.hs [hunk 1]
```diff
@@ -58,14 +58,11 @@ withJournalDoWeb :: WebOpts -> (WebOpts -> Journal -> IO a) -> IO a
...
 
-  -- https://github.com/simonmichael/hledger/issues/202
-  -- -f- gives [Error#yesod-core] <stdin>: hGetContents: illegal operation (handle is closed)
-  -- Also we may try to write to this file. Just disallow -.
-  when ("-" `elem` journalpaths) $  -- always non-empty
-    error' "hledger-web doesn't support -f -, please specify a file path"
-  mapM_ requireJournalFileExists journalpaths
+  -- If we're reading from a file, make sure it exists
+  forM_ journalpaths $ \path -> case path of
+    "-" -> return ()
```
[the remark this answers is not in the document]
Comment (zarybnicky) [reply hledger-web/Hledger/Web/Main.hs]: Yup, missed that.

File: hledger-web/Hledger/Web/Widget/Common.hs [hunk 2]
```diff
@@ -49,6 +50,7 @@ fromFormSuccess h (FormFailure _) = h
 fromFormSuccess _ (FormSuccess a) = pure a
 
 writeValidJournal :: MonadHandler m => FilePath -> Text -> m (Either String ())
```
[the remark this answers is not in the document]
Comment (zarybnicky) [reply hledger-web/Hledger/Web/Widget/Common.hs]: On revisiting this, I think that not printing out the journal is better, especially if we allow in-memory modifications of the journal.

**Comment by zarybnicky** [comment]
Sorry for the many force-pushes, I've been doing the changes one line at a time. I've rebased the commit and addressed the change requests, I think.

I'd like to implement in-memory parts of /edit and /upload as well, please don't merge this until then.

**Comment by zarybnicky** [comment]
Now also closes #950.

I've split the PR into two commits, one adding support for stdin to hledger-web, the other creating non-existent journals on startup. I've also added the missing /edit and /upload parts - now all of hledger-web supports stdin journals, and also filled in the missing `ensureJournalFileExists` line to hledger-ui's main.

**Comment by simonmichael** [comment]
Sorry for the delay on this. 
- Have you used the hledger-web stdin support a bit ? Working fine ?
- Why is the "create journal file if not exists" in this PR ? Are you sure that functionality is a good idea ?

**Comment by simonmichael** [comment]
> Why is the "create journal file if not exists" in this PR ? Are you sure that functionality is a good idea ?

Related: in what situations does hledger do that already ? I think only hledger add, right ?

**Comment by simonmichael** [comment]
Hi @zarybnicky, any thoughts on this ? It says I requested a change, I guess that was to remove/postpone the auto-create-missing-journal-file functionality.

**Comment by zarybnicky** [comment]
Oops, it seems I'd managed to archive the GitHub mail that has served as my reminder for this issue... I'll try to look it over tomorrow and see where it's at.

**Comment by simonmichael** [comment]
Ping

**Comment by zarybnicky** [comment]
Wow, I've been terrible regarding this PR. Real life stuff does that, sorry... I'll close this PR and open two separate ones instead

**Pull request closed without merging**